server/static/dashboard/localization-prototype.js

```diff
@@ -0,0 +1,200 @@
+    var newStyleSheet = document.createElement('style');
+
+    newStyles.forEach(function(newStyle) {
+        newStyleSheet.innerHTML += newStyle + '\n';
+    });
+
+    document.body.append(newStyleSheet);
+}
+
+
+// 4. Update diff behavior
+
+var lastCursorKerberos = undefined;
+var lastCursorTimestamp = undefined;
+var markers = {}; // file : line : message
+
+function updateMarkerDisplay() {
+  var errorElts = document.getElementsByClassName('diff-error');
+  for (var i = 0; i < errorElts.length; i++) {
+    errorElts[i].title = "";
+    errorElts[i].classList.remove('diff-error');
+  }
+  for (var fileName in markers) {
+    for (var lineNumber in markers[fileName]) {
+      var line = document.getElementById(fileName + '-' + lineNumber);
+      line.classList.add('diff-error');
+      line.title = markers[fileName][lineNumber][0];
+    }
+  }
+}
+
+if (queryMap['prototype'] == 'shrinkText') {
+    var oldUpdateDiff = updateDiff_basic;
+    updateDiff_basic = function(node, baseline, text, extraArgs) {
+        if (baseline === undefined || text === undefined) { return; }
+
+        var fileNode = node.parentNode.parentNode; // use to determine which file
+        var fileName = fileNode.children[0].innerHTML;
+
+        // track node of latest op
+        var op = extraArgs.op;
+        var count = 0;
+        var diffEltIndex = undefined;
+        var diffEltOffset = 0;
+        var opIndex = op === undefined ? -1 : (op[0].p[0] === 'text' ? -1 : op[0].oi[0]);
+
+        if (op !== undefined) {
+          if (op[0].p[0] === 'markers') {
+            var kerberos = op[0].p[1];
+            var errors = op[0].oi;
+            markers[fileName] = {};
+            for (var i = 0; i < errors.length; i++) {
+              var lineNumber = errors[i].lineNumber;
+              var message = errors[i].message;
+              if (markers[fileName][lineNumber] === undefined) {
+                markers[fileName][lineNumber] = [message];
+              } else {
+                markers[fileName][lineNumber].push(message);
+              }
+            }
+            updateMarkerDisplay();
+            return;
+          }
+          if (lastCursorKerberos === undefined) {
+            lastCursorKerberos = op[0].p[1];
+            lastCursorTimestamp = Date.now();
+          } else if (lastCursorKerberos === op[0].p[1]) {
+            lastCursorTimestamp = Date.now();
+          } else if (Date.now() - lastCursorTimestamp > 1000) {
+            lastCursorKerberos = op[0].p[1];
+            lastCursorTimestamp = Date.now();
+          } else {
+            opIndex = -1;
+          }
+        }
+
+        node.innerHTML = '';
+
+        var lineNumber = 0;
+        window.diff.diffLines(baseline.trim(), text.trim()).forEach(function(multipart) {
+
+          var parts = multipart.value.split('\n');
+          if (multipart.value[multipart.value.length-1] === '\n') {
+            parts = parts.slice(0, parts.length - 1);
+          }

```

Comment:
Shouldn't all the chunks (with possible exception of the last one, since with `trim`ed?) end with trailing newline? If so, simplify. If not, simplify and clarify.